### ISSUE TITLE
Proposal to make launcher use a fat jar of shaded sbt dependencies

### ISSUE THREAD

**Issue by olafurpg** [sbt/sbt#4183]
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/1.x/CONTRIBUTING.md) guidelines

I wrote up a proposal in https://github.com/olafurpg/sbt-big on how to speed up sbt launcher and provide more stable binary compatibility for the sbt plugin ecosystem. Related discussion https://github.com/sbt/zinc/issues/546 I've inlined the sbt-big readme below:

----

This project is a proposal to create a new module in sbt/sbt called `sbt-big`
that is a fat jar with all sbt dependencies shaded under an internal namespace.
By updating the sbt launcher to use `org.scala-sbt:sbt-big` instead of
`org.scala-sbt:sbt`, we gain

- faster startup time, by making the launcher download 1 fat jar containing all
  dependencies instead of resolving and downloading 71 jars for dependencies.
  This speedup should be particularly noticable in clean CI and docker environments and from locations with high latency such as Asia or Australia.
- more stable binary compatibility, by shading sbt dependencies under the
  `sbt.internal.*` package so that sbt plugins don't conflict with the sbt boot
  classpath.
  See [sbt/zinc#546](https://github.com/sbt/zinc/issues/546) for an example issue.
  By using `sbt-big`, the sbt project is safe to upgrade to the latest versions of its
  dependencies without breaking the sbt plugin ecosystem.

To try out `sbt-big` locally,

```
git clone https://github.com/olafurpg/sbt-big.git
cd sbt-big
sbt publishM2
```

What happens during publish is that:

- sbt-assembly includes the full dependency tree of sbt
```
$ du -h $HOME/.m2/repository/org/scala-sbt/sbt-big/1.1.0/sbt-big-1.1.0.jar
 21M    /Users/ollie/.m2/repository/org/scala-sbt/sbt-big/1.1.0/sbt-big-1.1.0.jar
```

- the published artifact contains 0 dependencies
```
$ cat $HOME/.m2/repository/org/scala-sbt/sbt-big/1.1.0/sbt-big-1.1.0.pom/sbt-big-1.1.0.pom | grep dependencies
<dependencies>
  <!--the dependency that was here has been absorbed via sbt-assembly-->
  <!--the dependency that was here has been absorbed via sbt-assembly-->
</dependencies>
```

- the fastparse dependency has been shaded into the `sbt.internal.fastparse` package
```
$ jar tf $HOME/.m2/repository/org/scala-sbt/sbt-big/1.1.0/sbt-big-1.1.0.jar | grep fastparse | head
sbt/internal/fastparse/
sbt/internal/fastparse/core/
sbt/internal/fastparse/parsers/
sbt/internal/fastparse/utils/
...
```

----

**Comment by CremboC** [comment]
Perhaps this fat jar can be then converted into a native binary use Graal's `native-image` for a further increase in startup time?

**Comment by olafurpg** [comment]
@CremboC I don't think Graal native images support the classloading APIs that sbt uses. Either way, Graal is unrelated to this proposal.

**Comment by eed3si9n** [comment]
### build.properties

One of the nice property of sbt is that we can up- or downgrade to 0.13.x and 1.x sbt versions by writing the sbt version number in `build.properties` file. This is possible because the launcher resolves `"org.scala-sbt" % "sbt" % x`. To keep this property, it might be worth considering keeping the continuity of the `"sbt"` module name, even if we adopt fat JAR.

### Metabuild reoslution

Similarly, sbt automatically injects `"org.scala-sbt" % "sbt" % x` to the metabuild.
A better solution for this actually would be to _exclude_ `"org.scala-sbt" % "sbt"` module in the metabuild instead (even when it gets pulled from plugins), and instead append sbt JAR(s) coming from the launcher.

**Comment by olafurpg** [comment]
@eed3si9n I agree with keeping the `"sbt"` module name, another concern by creating a separate `"sbt-big"` module is that the runtime classpath of sbt builds will be different depending on the version of the launcher. By keeping the module name, old launchers will behave the same as new launchers.

**Comment by andreaTP** [comment]
I took the time to benchmark the startup-time of sbt on this proposal against the default:
https://travis-ci.org/andreaTP/sbt-launcher-bench

this is a game changer, I really wanna see this option landing before `1.3.0` is out!

**Comment by eed3si9n** [comment]
#4431 is now merged but some tasks remain:

- [ ] `ivy.xml` needs to be post processed so `publishLocal` works.
- [ ] We need to add back the `libraryDependencies` that we are not shading.
- [ ] Flip the default to big by renaming "sbt-big" to "sbt", and "sbt" to something else.
- [ ] Publish X1, and test it on Community Build.

**Comment by olafurpg** [comment]
> We need to add back the `libraryDependencies` that we are not shading.

This change could remove the performance gains from shipping a fat jar. I personally care most about shading outdated libraries for a stable plugin ecosystem but the speedups are a nice side-effect.
